src/graphnet/components/loss_functions.py

```diff
@@ -31,6 +31,7 @@ def forward(
         self,
```

Comment:
In this module, I would suggest changing where the weights are multiplied from the within `_forward` method of the classes inheriting from `LossFunction` to the `forward` method, i.e., [here](https://github.com/icecube/graphnet/blob/main/src/graphnet/components/loss_functions.py#L48). If you change `LossFunction.forward` [here](https://github.com/icecube/graphnet/blob/main/src/graphnet/components/loss_functions.py#L30-L35) to have syntax `(self, prediction: Tensor, target: Tensor, weights: Optional[Tensor] = None, return_elements: bool = False)` and handle the default behaviour of `weights=None` to create a Tensor of all-ones, and then multiply these weights onto the output of `_forward`, you can leave all of the other classes unchanged. This is possible because `_forward` always return the per-sample losses. I think that would be a lot cleaner. And having a sensible default behaviour for `weight=None` implemented here means that you don't have to do this elsewhere, e.g. in the `Task` class.